core/store/src/trie/mod.rs

```diff
@@ -495,16 +500,18 @@ impl Trie {
         Some(PartialStorage { nodes: PartialState(nodes) })
     }
 
-    pub fn from_recorded_storage(partial_storage: PartialStorage) -> Self {
+    pub fn from_recorded_storage(partial_storage: PartialStorage, root: StateRoot) -> Self {
         let recorded_storage =
             partial_storage.nodes.0.into_iter().map(|value| (hash(&value), value)).collect();
-        Trie {
-            storage: Box::new(TrieMemoryPartialStorage {
-                recorded_storage,
-                visited_nodes: Default::default(),
-            }),
-            flat_state: None,
-        }
+        let storage = Box::new(TrieMemoryPartialStorage {
+            recorded_storage,
+            visited_nodes: Default::default(),
+        });
+        Self::new(storage, root, None)
+    }
+
+    pub fn get_root(&self) -> &StateRoot {
```

Comment:
```suggestion
    pub fn root(&self) -> &StateRoot { 
        &self.root
    }
```

would be more idiomatic: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter. As this is a new API, I'd maybe stick to API guidelines.

Also, maybe `fn state_root() -> &StateRoot` ? To make it unambigious with root node.

Reply:
Ah, I see that's not exactly the new API, we keep it consistent with `StoreUpdate`. Better leave it as is, and introduce TrieAccess trait first!